File: solutions/_1496_path_crossing/solution.cpp
#include <set>
#include <string>
using namespace std;

class Solution
{
public:
    bool isPathCrossing(string path)
    {
        int x = 0, y = 0;
        set<pair<int, int>> visited;
        visited.insert({x, y});
        for (char step : path)
        {
            switch (step)
            {
            case 'N':
                y++;
                break;
            case 'S':
                y--;
                break;
            case 'E':
                x++;
                break;
            case 'W':
                x--;
                break;
            }

            if (visited.find({x, y}) != visited.end())
                return true;
            visited.insert({x, y});
        }
        return false;
    }
};

/*
https://leetcode.com/problems/path-crossing/
*/

File: solutions/_1496_path_crossing/solution.py
class Solution:
    def isPathCrossing(self, path: str) -> bool:
        visited = set()
        x, y = 0, 0
        visited.add((x, y))
        for step in path:
            if step == "N":
                y += 1
            elif step == "S":
                y -= 1
            elif step == "E":
                x += 1
            elif step == "W":
                x -= 1
            if (x, y) in visited:
                return True
            visited.add((x, y))

        return False


# https://leetcode.com/problems/path-crossing/
